server/src/main/java/org/elasticsearch/common/rounding/Rounding.java

```diff
@@ -125,44 +125,79 @@ public byte id() {
+
+            return utcMillis;
+        }
+
         @Override
         public long round(long utcMillis) {
-            long rounded = field.roundFloor(utcMillis);
-            if (timeZone.isFixed() == false) {
-                // special cases for non-fixed time zones with dst transitions
-                if (timeZone.getOffset(utcMillis) != timeZone.getOffset(rounded)) {
-                    /*
-                     * the offset change indicates a dst transition. In some
-                     * edge cases this will result in a value that is not a
-                     * rounded value before the transition. We round again to
-                     * make sure we really return a rounded value. This will
-                     * have no effect in cases where we already had a valid
-                     * rounded value
-                     */
-                    rounded = field.roundFloor(rounded);
-                } else {
-                    /*
-                     * check if the current time instant is at a start of a DST
-                     * overlap by comparing the offset of the instant and the
-                     * previous millisecond. We want to detect negative offset
-                     * changes that result in an overlap
-                     */
-                    if (timeZone.getOffset(rounded) < timeZone.getOffset(rounded - 1)) {
-                        /*
-                         * we are rounding a date just after a DST overlap. if
-                         * the overlap is smaller than the time unit we are
-                         * rounding to, we want to add the overlapping part to
-                         * the following rounding interval
-                         */
-                        long previousRounded = field.roundFloor(rounded - 1);
-                        if (rounded - previousRounded < field.getDurationField().getUnitMillis()) {
-                            rounded = previousRounded;
-                        }
+
+            // field.roundFloor() works as long as the offset doesn't change.  It is worth getting this case out of the way first, as
+            // the calculations for fixing things near to offset changes are a little expensive and are unnecessary in the common case
+            // of working in UTC.
+            if (timeZone.isFixed()) {
+                return field.roundFloor(utcMillis);
+            }
+
+            // When rounding to hours we consider any local time of the form 'xx:00:00' as rounded, even though this gives duplicate
+            // bucket names for the times when the clocks go back. Shorter units behave similarly. However, longer units round down to
+            // midnight, and on the days where there are two midnights we would rather pick the earlier one, so that buckets are
+            // uniquely identified by the date.
+            boolean roundingToMidnight = field.getDurationField().getUnitMillis() > 60L * 60L * 1000L;
```

Comment:
Since the field shouldn't change (maybe it should final?), it looks like this could be computed once, e.g. in the ctor. Maybe then it would make sense to call this flag something different, maybe `dayUnitOrBigger`, or `unitRoundsToMidnight`

Reply:
Makes sense, I pushed 8c4b99c5b4 to do this. It can't be `final` because it's set in the `readFrom()` method.